In [ ]:
# Составление "мешка слов"

reviews = [
    '''Брала из-за переломанного провода в старых наушниках, а в этих плетеный провод, долговечней должен быть. К телефону подключаю через провод otg. (покупала отдельно). Слушаю музыку только в режиме "игры", на других плосковат звук почему то а на этом режиме басы объемней всего звучат. Режим 7.1 словно делает громче какие-то звуки, в играх может быть нормально, но для прослушивания музыки звучит неестественно.
    По поводу удобства. Начитавшись комментариев, надеялась, что будут жать, супер плотно прилегать, а нет. (размер головы от левого уха через затылок до правого: 26.5 см) Может у меня голова маленькая, или производитель начал делать ослабленные варианты, но для меня даже слабовато прилегают, хотелось бы посильнее. Но на голове сидят, при повороте не слетают.
    Может быть их можно в домашних условиях согнуть посильнее в ободке, но я боюсь сломать.
    Подсветка реально яркая, ночью если спать с музыкой, надеть лучше маску для сна на глаза.
    В целом, отличные наушники, полностью соответствуют заявленным функциям. Цвет приятный, все детали качественные, собрано хорошо. Провод и пульт легкие, не понимаю тех кто писал про тяжесть, мне отлично.''',
    
    'наушники огонь🔥',

    'звук очень резкий и не оптимизированный',
    
    'Пользуюсь наушниками 4 месяц все прекрасно работает ,но иногда бывает что микро надо чуть подвернуть что бы снова заработал но это очень редкая ситуация .Микрофон работает без каких либо проблем ,звук остаётся четкий и хороший.Единственный минус от них у меня уши болят после 2-3 часов',
    
    'наушники отличные, купили на НГ ребёнку, хотел такие. проверила, всё работает, звук устраивает. сначала думала микрофон не работает, написала вопрос продавцу, ответ ещё не поступил, но проблема решена) оказалось микрофон плохо воткнула, не до конца. имейте ввиду, что воткнуть надо сильнее)) подсветка есть, переключатель работает.',
    
    'всё есть мешок сами наушники и бумажка, соответствует характеристикам всё работает.',
    
    'наушники суперские.рекомендую 👍',
    
    'Пришли быстро. Коробка была немного помята, сама коробка запломбирована. Наушники очень лёгкие покупал в подарок, надеюсь очень понравится',
    
    'Хорошие наушники, достаточно длинный кабель. Звук нормальный, иногда появляется какой-то шум, но не часто.',
    
    'прошло около 3 недель,всё было отлично,ни к чему не придраться,но с недавнего микрофон вовсе не работает. сначало звук был как будто я разговариваю из далека комнаты,это если повысить голос , через некоторое время микрофон совсем перестал работать,не понимаю в чём дело...хотел избежать покупку отдельного микрофона,так как знал что у фифайн наушниках хороший микрофон, но видимо мне не повезло,жаль.',
    
    'товар соответствует описанию',
    
    'Почти сразу начал отключаться правый наушник. Первое время лечилось переподключением в разъем, в интернете подобной проблемы и ее решения не встретил. Максимум писали про шум/шипение в одном ухе. Сейчас и переподключение не спасает.',
    
    'наушники просто супер, но вот качество звука не совсем приятное',
    
    'брак',
    
    'После использования уши болят ужасно плохо сделаны амбушуры лучше подкопить и купить получше',
    
    'Отличные наушники все работает',
    
    'Да вообще бомбово,микрофон фонит сами наушники трещат ,я рад покупке!!!!!!!!!!!!!!!!!!!!!!!!!!!',
    
    'По началу качество хорошее, со звуком проблем не было, но через пол года начались непонятные похрипывания в наушниках, кнопка выключения микрофона туговата, провод держит ту форму, которая пол года назад была в упаковке, распрямляться не желает.',
    
    'Сборка какая-то некачественная. Это мои вторые наушники от этого бренда, и в первых она была получше',
    
    'хорошие наушники',
    
    'хороший звук, не плохой микрофон. коробка пришла чуть помятая, но не критично. ещё немного давят на голову спустя два часа использования, но это должно пройти через неделю другую',
    
    'Отлично',
    
    'Так себе наушники, микрофон - 2/5, качество звука в самих наушниках - 2/5, постоянный микро-шум, не рекомендую к покупке, проще переплатить и взять лучше , чем брать такой товар',
    
    'Хорошие наушники с прекрасным звучанием и хорошим микрофоном',

    'Хорошие, удобные наушники, с неплохим микрофоном, за такую цену более чем.',
]

print(f'Количество отзывов: {len(reviews)}')

Количество отзывов: 25


In [5]:
# Выявление наиболее популярных слов, чтобы исключить из них возможные стоп-слова
words_dict = {}

for review in reviews:
    for word in review.split():
        word = word.lower()
        if word in words_dict:
            words_dict[word] += 1
        else:
            words_dict[word] = 1
            
sorted_words = sorted(words_dict.items(), key=lambda item: item[1], reverse=True)

for word, count in sorted_words:
    print(f"{word}: {count}")

не: 18
в: 16
но: 12
наушники: 11
и: 10
на: 6
через: 5
звук: 5
что: 5
микрофон: 5
с: 4
наушники,: 4
очень: 4
а: 3
к: 3
провод: 3
может: 3
для: 3
от: 3
у: 3
меня: 3
лучше: 3
все: 3
работает: 3
это: 3
всё: 3
работает.: 3
коробка: 3
была: 3
хорошие: 3
качество: 3
наушниках,: 2
режиме: 2
быть: 2
по: 2
до: 2
начал: 2
бы: 2
я: 2
подсветка: 2
если: 2
отличные: 2
понимаю: 2
про: 2
мне: 2
иногда: 2
надо: 2
чуть: 2
проблем: 2
уши: 2
болят: 2
после: 2
работает,: 2
ещё: 2
плохо: 2
сами: 2
соответствует: 2
немного: 2
как: 2
,: 2
время: 2
совсем: 2
наушниках: 2
хороший: 2
товар: 2
звука: 2
получше: 2
пол: 2
года: 2
-: 2
2/5,: 2
брала: 1
из-за: 1
переломанного: 1
провода: 1
старых: 1
этих: 1
плетеный: 1
провод,: 1
долговечней: 1
должен: 1
быть.: 1
телефону: 1
подключаю: 1
otg.: 1
(покупала: 1
отдельно).: 1
слушаю: 1
музыку: 1
только: 1
"игры",: 1
других: 1
плосковат: 1
почему: 1
то: 1
этом: 1
басы: 1
объемней: 1
всего: 1
звучат.: 1
режим: 1
7.1: 1
словно: 1
делает: 1
громче: 1
какие-то: 1
звуки,: 1
иг

In [7]:
# Составление словаря стоп-слов
stop_words = [
    "не", "в", "но", "и", "на", "через", "что", "с", "а", "для", "от", 
    "у", "меня", "все", "это", "как", "по", "если", "про", "мне", "иногда", 
    "надо", "чуть", "после", "ещё", "плохо", "сами", "чуть", "какие-то", 
    "или", "есть", "всё", "при", "в", "их", "можно", "при", "до", "был", 
    "что", "везде", "где", "немного", "так", "эти", "все"
]

print(f'Количество стоп-слов: {len(stop_words)}')

Количество стоп-слов: 46


In [ ]:
# Масштабирование данных с помощью tf-idf
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer(stop_words=stop_words)
X_tfidf = tfidf_vectorizer.fit_transform(reviews)

print("TF-IDF матрица:", X_tfidf.toarray())

TF-IDF матрица: [[0.08298567 0.08298567 0.         ... 0.         0.08298567 0.08298567]
 [0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.         0.        ]
 ...
 [0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.         0.        ]]


c:\Users\bread\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\feature_extraction\text.py:406: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['какие', 'то'] not in stop_words.
  warnings.warn(


In [ ]:
# Использование n-gram (1, 2, 3), pipeline(vectorizer, regressor) 
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# Пример меток для отзывов: 0 - негативный, 1 - положительный
labels = [1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1]

X_train, X_test, y_train, y_test = train_test_split(reviews, labels, test_size=0.2, random_state=42)

pipeline = Pipeline([
    ('vectorizer', TfidfVectorizer(ngram_range=(1, 2), stop_words=stop_words)),
    ('classifier', LogisticRegression())
])

pipeline.fit(X_train, y_train)

y_pred = pipeline.predict(X_test)
print("Точность модели:", accuracy_score(y_test, y_pred))


Точность модели: 0.6


c:\Users\bread\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\feature_extraction\text.py:406: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['какие', 'то'] not in stop_words.
  warnings.warn(


In [24]:
# Исследование метрик модели
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

print("Точность модели:", accuracy_score(y_test, y_pred))

print("\nОтчет по меткам (precision, recall, f1-score):")
print(classification_report(y_test, y_pred))

print("\nМатрица ошибок:")
print(confusion_matrix(y_test, y_pred))

print("\nРеальные метки vs предсказанные метки:")
for true, pred in zip(y_test, y_pred):
    print(f"Реальная метка: {true}, Предсказанная метка: {pred}")

Точность модели: 0.6

Отчет по меткам (precision, recall, f1-score):
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         2
           1       0.60      1.00      0.75         3

    accuracy                           0.60         5
   macro avg       0.30      0.50      0.38         5
weighted avg       0.36      0.60      0.45         5


Матрица ошибок:
[[0 2]
 [0 3]]

Реальные метки vs предсказанные метки:
Реальная метка: 1, Предсказанная метка: 1
Реальная метка: 0, Предсказанная метка: 1
Реальная метка: 1, Предсказанная метка: 1
Реальная метка: 1, Предсказанная метка: 1
Реальная метка: 0, Предсказанная метка: 1


c:\Users\bread\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\bread\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\bread\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, mo

In [23]:
# Вывод слов с наибольшим вкладом
model = pipeline.named_steps['classifier']
coefficients = model.coef_

words = pipeline.named_steps['vectorizer'].get_feature_names_out()
top_10_words = sorted(zip(coefficients[0], words), reverse=True)[:10]
print("Топ-10 слов с наибольшим вкладом:")

for word in top_10_words:
    print(word)

Топ-10 слов с наибольшим вкладом:
(0.37233526258975597, 'наушники')
(0.20345488988714472, 'отлично')
(0.20044232341912613, 'работает')
(0.19388829336877128, 'хорошие')
(0.17013451476980007, 'соответствует')
(0.16310570421985865, 'хорошие наушники')
(0.15819196613134, 'огонь')
(0.15819196613134, 'наушники огонь')
(0.13564627339456928, 'отличные')
(0.12360145011512143, 'отличные наушники')


## **Лабораторная работа №7**
### **Вывод:**

Для данной лабораторной работы в качестве **"мешка слов"** использовался список отзывов на наушники.

#### Результаты

---

Точность модели: 0.6

Отчет по меткам (precision, recall, f1-score):
              
                 precision    recall  f1-score   support
           0       0.00      0.00      0.00         2
           1       0.60      1.00      0.75         3

    accuracy                           0.60         5
    macro avg       0.30      0.50     0.38         5
    weighted avg    0.36      0.60     0.45         5

Модель показала не самую лучшую точность, скорее всего из-за малого количества данных.